fvm/environment/uuids.go

```diff
@@ -39,45 +50,118 @@ func (generator ParseRestrictedUUIDGenerator) GenerateUUID() (uint64, error) {
 
 type uUIDGenerator struct {
 	tracer tracing.TracerSpan
+	log    zerolog.Logger
 	meter  Meter
 
 	txnState state.NestedTransactionPreparer
+
+	blockHeader *flow.Header
+	txnIndex    uint32
+
+	initialized bool
+	partition   byte
+	registerId  flow.RegisterID
+}
+
+func uuidPartition(blockId flow.Identifier, txnIndex uint32) byte {
+	// The deterministically random offset is used to ensure the partitioned
+	// counters are roughly balanced.  Any byte in the sha hash is sufficiently
+	// random/uniform for this purpose.
+	partitionOffset := sha256.Sum256(blockId[:])[0]
```

Comment:
if two blocks have the same partition offset, then the first uuid from their first transaction would be the same, right? 

Is that OK for two transactions from different blocks to have the same UUID? Or should we call it BUID (block-wise unique ID) or something else

Reply:
just to be clear, transaction conflict never causes correctness issue (assuming i implemented mvcc correctly).  it only hurts performance due to the need to retry the transaction.

Reply:
I think I know where I was confused. 

I thought the uuid is only created within unique within a block, and won't be saved to storage. 

But actually, we want to save it to storage, so that future transactions will be guaranteed to get a different uuid by incrementing it. 

So overtime, the uuid will keep incrementing as the transactions are requesting uuid values continuously, and one day eventually it will get close to the limit, which MaxUint56 value for each partitioned uuid. 

I think it's better to explain in the comment that UUID is implemented by storing a counter in storage so that whenever a transaction requesting it, it will get incremented. And the key for the counter value is partitioned so that multiple transactions can execute concurrently without getting the same counter value from the storage. In the case where concurrently executed transaction did get the same counter value from reading it from the same partition, the transaction coordinator will re-execute the later transaction which will get read the correct counter value for UUID.